src/poly/domain.rs

```diff
@@ -268,6 +252,102 @@ impl<G: Group> EvaluationDomain<G> {
         }
     }
 
+    fn op_extended(
+        &self,
+        mut left: Polynomial<G, ExtendedLagrangeCoeff>,
+        right: &Polynomial<G, ExtendedLagrangeCoeff>,
+        rotation: Rotation,
+        op: impl Fn(&mut G, &G) + Send + Sync + 'static,
+    ) -> Polynomial<G, ExtendedLagrangeCoeff>
+    where
+        G: Field,
+    {
+        // Rotation while in the extended domain is simply exaggerated by how
+        // much larger the extended domain is compared to the original domain.
+        let rotation = (1 << (self.extended_k - self.k)) * rotation.0;
+
+        parallelize(&mut left.values, |lhs, start| {
+            let start = ((((start + self.extended_len()) as i32) + rotation) as usize)
+                % self.extended_len();
+
+            for (lhs, rhs) in lhs
+                .iter_mut()
+                .zip(right.values[start..].iter().chain(right.values.iter()))
+            {
+                op(lhs, rhs);
```

Comment:
This does a dynamic function call for every element of the extended domain, which may not be a negligible cost. Consider inlining into `{mul,add,sub}_extended`, or using a macro.

Reply:
It's not a dynamic function call. `op` is of type `F` bounded by `F: Fn(&mut G, &G) + Send + Sync + 'static` which undergoes monomorphization. It would be dynamic dispatch if `op` was `dyn`.